File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/container/slots/TypeSlot.java
package generations.gg.generations.core.generationscore.common.world.container.slots;

import com.mojang.datafixers.util.Pair;
import dev.architectury.registry.registries.RegistrySupplier;
import earth.terrarium.common_storage_lib.resources.item.ItemResource;
import earth.terrarium.common_storage_lib.storage.base.CommonStorage;
import earth.terrarium.common_storage_lib.storage.util.MenuStorageSlot;
import generations.gg.generations.core.generationscore.common.GenerationsCore;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.Container;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.DoubleSupplier;

import static net.minecraft.world.inventory.InventoryMenu.BLOCK_ATLAS;

public class TypeSlot extends MenuStorageSlot {
    private final RegistrySupplier<Item> candy;
    private final DoubleSupplier supplier;
    private final Pair<ResourceLocation, ResourceLocation> pair;
    public int originalX;
    public int originalY;

    public TypeSlot(CommonStorage<ItemResource> itemHandler, int index, int xPosition, int yPosition, RegistrySupplier<Item> candy, String elementName, DoubleSupplier supplier) {
        super(itemHandler, index, xPosition, yPosition);
        this.candy = candy;
        this.supplier = supplier;
        pair = new Pair<>(BLOCK_ATLAS, GenerationsCore.id("type_slots/" + elementName));
    }

    public static void interpolateVectors(TypeSlot slot, int x1, int y1, int x2, int y2, double t) {
        slot.x = Math.round((float) (slot.originalX + (x2 - slot.originalX) * t));
        slot.y = Math.round((float) (slot.originalY + (y2 - slot.originalY) * t));
    }

    @Override
    public boolean isActive() {
        return supplier.getAsDouble() < 1.0f;
    }

    @Override
    public boolean mayPlace(@NotNull ItemStack stack) {
        return stack.is(candy.get());
    }

    @Nullable
    @Override
    public Pair<ResourceLocation, ResourceLocation> getNoItemIcon() {
        return pair;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/container/slots/TypeSlot.kt
package generations.gg.generations.core.generationscore.common.world.container.slots

import com.mojang.datafixers.util.Pair
import generations.gg.generations.core.generationscore.common.GenerationsCore.id
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.Container
import net.minecraft.world.inventory.InventoryMenu
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import java.util.function.DoubleSupplier

//Keeping around even with the defunct machine block because you never know when a type themeed slot would be useful
class TypeSlot(
    itemHandler: Container,
    index: Int,
    xPosition: Int,
    yPosition: Int,
    private val candy: Item,
    elementName: String,
    private val supplier: DoubleSupplier
) :
    Slot(itemHandler, index, xPosition, yPosition) {
    private val pair =
        Pair(
            InventoryMenu.BLOCK_ATLAS, id(
                "type_slots/$elementName"
            )
        )
    var originalX: Int = 0
    var originalY: Int = 0

    override fun isActive(): Boolean {
        return supplier.asDouble < 1.0f
    }

    override fun mayPlace(stack: ItemStack): Boolean {
        return stack.`is`(candy)
    }

    override fun getNoItemIcon(): Pair<ResourceLocation, ResourceLocation>? {
        return pair
    }

    companion object {
        @JvmStatic
        fun interpolateVectors(slot: TypeSlot, x1: Int, y1: Int, x2: Int, y2: Int, t: Double) {
            slot.x = Math.round((slot.originalX + (x2 - slot.originalX) * t).toFloat())
            slot.y = Math.round((slot.originalY + (y2 - slot.originalY) * t).toFloat())
        }
    }
}
